dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs

```diff
@@ -198,6 +225,119 @@ public static class OpenApiKernelExtensions
+
+        if (!File.Exists(apiManifestFilePath))
+        {
+            throw new FileNotFoundException($"ApiManifest file not found: {apiManifestFilePath}");
+        }
+
+        string apiManifestFileJsonContents = File.ReadAllText(apiManifestFilePath);
+        JsonDocument jsonDocument = JsonDocument.Parse(apiManifestFileJsonContents);
+
+        ApiManifestDocument document = ApiManifestDocument.Load(jsonDocument.RootElement);
+
+        using var openApiReadinghttpClient = new HttpClient();
+
+        var functions = new List<KernelFunction>();
+        foreach (var apiDependency in document.ApiDependencies)
+        {
+            var apiName = apiDependency.Key;
+            var apiDependencyDetails = apiDependency.Value;
+
+            var apiDescriptionUrl = apiDependencyDetails.ApiDescriptionUrl;
+
+            using var stream = await openApiReadinghttpClient.GetStreamAsync(new Uri(apiDescriptionUrl)).ConfigureAwait(false);
+
+            var parseResult = await new OpenApiStreamReader(new()
+            {
+                BaseUrl = new(apiDescriptionUrl)
+            }
+            ).ReadAsync(stream, cancellationToken).ConfigureAwait(false);
+            var openApiDocument = parseResult.OpenApiDocument;
+
+            var requestUrls = new Dictionary<string, List<string>>();
+            var paths = apiDependencyDetails.Requests.Select(request => request.UriTemplate);
+            foreach (var path in paths)
+            {
+                if (path is null)
+                {
+                    continue;
+                }
+
+                requestUrls.Add(path, new List<string> { "GET" });
+            }
+
+            var predicate = OpenApiFilterService.CreatePredicate(null, null, requestUrls, openApiDocument);
+            var filteredOpenApiDocument = OpenApiFilterService.CreateFilteredDocument(openApiDocument, predicate);
+            ILoggerFactory loggerFactory = kernel.LoggerFactory;
+            var logger = loggerFactory.CreateLogger(typeof(OpenApiKernelExtensions));
+
+            var serverUrl = filteredOpenApiDocument.Servers.FirstOrDefault()?.Url;
+
+            var runner = new RestApiOperationRunner(
+                httpClient,
+                executionParameters?.AuthCallback,
+                executionParameters?.UserAgent,
+                executionParameters?.EnableDynamicPayload ?? false,
+                executionParameters?.EnablePayloadNamespacing ?? false);
+
+            foreach (var path in filteredOpenApiDocument.Paths)
+            {
+                var operations = OpenApiDocumentParser.CreateRestApiOperations(serverUrl, path.Key, path.Value);
+                foreach (RestApiOperation operation in operations)
+                {
+                    try
+                    {
+                        logger.LogTrace("Registering Rest function {0}.{1}", pluginName, operation.Id);
+
+                        static string GenerateFunctionName(RestApiOperation operation)
```

Comment:
I did not know that the operationId property was optional. Thank you for pointing that out. Later, we need to consolidate our approach to naming so that all functions have the same naming pattern, regardless of how they are imported. Perhaps we could use a fallback logic: use the operationId as a name if provided; otherwise, fall back to a custom format, such as operation.Method + operation.Path.

Reply:
I gave this a try and ended up not being able to import functions because there is a `.` in our operationIds in Graph OpenAPI docs. I was able to bypass that error after changing the skipped `KernelException` to `ArgumentException` because that's what's been thrown in `ValidFunctionName` method.

https://github.com/microsoft/semantic-kernel/blob/f2fb71d9d14fa6fd3eefd518d7ae3bb55eef5d16/dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs#L359
 
One concern I have with the fallback is that we may break the uniqueness constraint by sanitizing two different operations into colliding names. For example, `me.ListMessages` and `me-ListMessages` would sanitize to the same `meListMessages`. What is the likelihood of that? Pretty low, but it is something to consider.

Given that operationIds must be unique as long as they are provided, can we use the same valid character set for the function name regex?

There is also a chance to have collision between an operationId from one operation and a fallback coming from another operation. Again, low likelihood, but possible.

I feel that this deserves to have its own GitHub Issue.

Reply:
looks like there was an existing issue regarding the skipped exception: https://github.com/microsoft/semantic-kernel/issues/4567

Reply:
This is the fix for the issue - https://github.com/microsoft/semantic-kernel/pull/4708. 

> I feel that this deserves to have its own GitHub Issue.

Agreed, let's address this in a separate issue. For now, I would suggest keeping the current approach for names, which assumes the use of sanitized operation IDs (using the ConvertOperationIdToValidFunctionName method) as the operation/function name. We can consider changing this later when we have more data/information.